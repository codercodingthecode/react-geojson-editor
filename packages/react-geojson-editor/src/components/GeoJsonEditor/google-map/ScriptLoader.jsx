/* global google */
/* eslint-disable react/no-did-mount-set-state */
import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { fetchJsScript } from '../../../utils';

// withScriptjs
const LOADING_STATE_NONE = 'NONE';
const LOADING_STATE_BEGIN = 'BEGIN';
const LOADING_STATE_LOADED = 'LOADED';

// context is the global window.google object set when the script has been loaded
export const ScriptContext = React.createContext(null);

export class ScriptLoader extends Component {
  constructor(props) {
    super(props);
    this.handleLoaded = this.handleLoaded.bind(this);
  }
    state = {
      loadingState: LOADING_STATE_NONE,
      google: null,
    };

    componentDidMount() {
      const { scriptUrl } = this.props;
      const { loadingState } = this.state;
      if (loadingState !== LOADING_STATE_NONE) return;
      this.setState({ loadingState: LOADING_STATE_BEGIN }, () => {
        // @TODO have fetchJsScript provide google object in the callback
        // fetchJsScript(scriptUrl, google => this.handleLoaded(google))
        fetchJsScript(scriptUrl, () => this.handleLoaded(google));
      });
    }
    componentWillUnmount() {
      this.isUnmounted = true;
    }
    isUnmounted = false

    handleLoaded(google) {
      if (this.isUnmounted) {
        return;
      }
      this.setState({
        google,
        loadingState: LOADING_STATE_LOADED,
      });
    }

    render() {
      const { google, loadingState } = this.state;
      if (loadingState !== LOADING_STATE_LOADED) { return <div> Loading ... </div>; }

      if (!google) { return <div> Error Loading Script </div>; }

      return (
        <ScriptContext.Provider value={google}>
          {this.props.children}
        </ScriptContext.Provider>
      );
    }
}

ScriptLoader.propTypes = {
  children: PropTypes.node.isRequired,
  scriptUrl: PropTypes.string.isRequired,
};

export default ScriptLoader;
