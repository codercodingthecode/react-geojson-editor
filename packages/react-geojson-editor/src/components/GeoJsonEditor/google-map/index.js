export { ScriptLoader } from './ScriptLoader';
export { GoogleMapInitialzer, GoogleMapContext } from './GoogleMapInitialzer';
export { GoogleMap, GoogleMapWithLoader } from './GoogleMap';
export { PolygonDataProvider, PolygonDataContext } from './PolygonDataProvider';
