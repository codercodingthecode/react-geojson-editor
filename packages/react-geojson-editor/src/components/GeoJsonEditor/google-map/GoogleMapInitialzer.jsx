/* eslint-disable react/no-unused-state */
/* eslint-disable react/no-did-mount-set-state */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { ScriptContext } from './ScriptLoader';

// withGoogleMap
export const GoogleMapContext = React.createContext({
  google: null,
  map: null,
});
export class GoogleMapInitialzer extends Component {
    static contextType = ScriptContext;

    constructor(props, context) {
      super(props, context);
      if (!context) { throw new Error('No google object found, please make sure you have loaded google map script'); }
      this.mapElement = React.createRef();
    }
    state = { google: null, map: null };

    componentDidMount() {
      const { center, zoom } = this.props;
      const google = this.context;
      const map = new google.maps.Map(this.mapElement.current, {
        center,
        zoom,
        fullscreenControl: false,
        streetViewControl: false,
        mapTypeControlOptions: {
          style: google.maps.MapTypeControlStyle.DROPDOWN_MENU,
        },
      });
      this.setState({ google, map });
    }

    componentDidUpdate(prevProps) {
      const { center } = this.props;
      if (prevProps.center !== center) { this.state.map.setCenter(center); }
    }

    render() {
      const { width, height } = this.props;
      const { map } = this.state;
      return (
        <div style={{ width, height }} ref={this.mapElement}>
          {
            map ?
              <GoogleMapContext.Provider value={this.state}>
                {this.props.children}
              </GoogleMapContext.Provider> :
              null
          }
        </div>
      );
    }
}

GoogleMapInitialzer.propTypes = {
  center: PropTypes.shape({
    lat: PropTypes.number,
    lng: PropTypes.number,
  }).isRequired,
  children: PropTypes.node.isRequired,
  // @note height needs to be greater than 0px
  // @source: https://developers.google.com/maps/documentation/javascript/adding-a-google-map
  height: PropTypes.number.isRequired,
  width: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]).isRequired,
  zoom: PropTypes.number.isRequired,
};

export default GoogleMapInitialzer;
