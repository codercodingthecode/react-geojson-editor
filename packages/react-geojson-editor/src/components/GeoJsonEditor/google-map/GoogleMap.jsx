// 1. has a Context
// 2. has a wrapper component for
//    providing this context to child components
//      (ex: allowing DrawingManager component to access map instance)

import React from 'react';
import PropTypes from 'prop-types';

// import { GoogleMapContext } from './GoogleMapInitialzer';
import { ScriptLoader, ScriptContext } from './ScriptLoader';
import { GoogleMapInitialzer } from './GoogleMapInitialzer';

// Google namespace
// @type {{ maps: { [key:string]: any }}} google

export function GoogleMapWithLoader({
  googleMapURL, children, center, zoom, height, width,
}) {
  return (
    <ScriptLoader scriptUrl={googleMapURL}>
      <GoogleMapInitialzer
        center={center}
        zoom={zoom}
        height={height}
        width={width}
      >
        {children}
      </GoogleMapInitialzer>
    </ScriptLoader>
  );
}

/**
 * provide child components, with access to map instance
 * @param {} props
 */
export function GoogleMap({
  google, children, center, zoom, height, width,
}) {
  // in react-google-maps, this where they add event listeners
  return (
    <ScriptContext.Provider value={google}>
      <GoogleMapInitialzer
        center={center}
        zoom={zoom}
        height={height}
        width={width}
      >
        {children}
      </GoogleMapInitialzer>
    </ScriptContext.Provider>
  );
}
export default GoogleMap;

GoogleMapWithLoader.propTypes = {
  googleMapURL: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
  center: PropTypes.shape({
    lat: PropTypes.number,
    lng: PropTypes.number,
  }).isRequired,
  height: PropTypes.number.isRequired,
  width: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  zoom: PropTypes.number,
};

GoogleMap.propTypes = {
  google: PropTypes.object.isRequired, // eslint-disable-line react/forbid-prop-types
  children: PropTypes.node.isRequired,
  center: PropTypes.shape({
    lat: PropTypes.number,
    lng: PropTypes.number,
  }).isRequired,
  height: PropTypes.number.isRequired,
  width: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  zoom: PropTypes.number,
};

const GoogleMapDefaultProps = {
  width: '100%',
  zoom: 5,
};

GoogleMapWithLoader.defaultProps = GoogleMapDefaultProps;
GoogleMap.defaultProps = GoogleMapDefaultProps;
