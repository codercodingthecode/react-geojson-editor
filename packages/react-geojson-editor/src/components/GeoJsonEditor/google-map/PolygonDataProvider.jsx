/* eslint-disable react/no-did-update-set-state */
/* eslint-disable react/no-unused-state */

import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { subtractPolygons } from '../../../turfjs-operations/subtraction';
import {
  convertPolygonToGeoJsonMulti, convertGeojsonToPolygon,
  convertGeojsonToGMData, gmDataToPolygonList,
  createGeoJson,
} from '../../../gm-operations/convert-polygon';
import { GeoJsonType } from '../../../custom-types';
import { GoogleMapContext } from './GoogleMapInitialzer';

/**
 * In charge of keeping track of polygons
 * @NOTE PolygonDataProvider is not allowed to call polygon methods (
 *      ex: polygon.setEditable or polygon.setMap)
 *      these calls should done outside
 */
export const PolygonDataContext = React.createContext({
  gmapData: null, // of type google.maps.Data, used to present existing polygons
  newPolygonList: [], // of type google.maps.Polygon
  polygonList: [], // converted from gmapData for editing purposes
  selectedPolygonList: [], // selected from either newPolygonList or polygonList

  // Actions
  // createGmapData: null, // of type Function
  addNewPolygon: null,
  // removeNewPolygon: null,
  selectPolygon: null,
  unSelectPolygon: null,

  // gmapData to polygonList
  // generatePolygonList: null,

  // Editor actions (applied on selected polygon) ??
  addAction: null, // combine selected polygons
  subtractAction: null,
  deleteAction: null,
  centerAction: null,

  discardChanges: null,
  persistChanges: null,
});
export class PolygonDataProvider extends Component {
      static contextType = GoogleMapContext;
      constructor(props, context) {
        super(props, context);
        // this.createGmapData = this.createGmapData.bind(this);
        this.addNewPolygon = this.addNewPolygon.bind(this);
        this.selectPolygon = this.selectPolygon.bind(this);
        this.unSelectPolygon = this.unSelectPolygon.bind(this);
        this.subtractAction = this.subtractAction.bind(this);
        this.deleteAction = this.deleteAction.bind(this);

        this.discardChanges = this.discardChanges.bind(this);
        this.persistChanges = this.persistChanges.bind(this);
        // this.generatePolygonList = this.generatePolygonList.bind(this);

        const gmapData = props.initialGeojson ?
          convertGeojsonToGMData(props.initialGeojson, context) :
          null;
        this.state = {
          gmapData,
          newPolygonList: [],
          polygonList: gmapData ? gmDataToPolygonList(gmapData, context) : [],
          selectedPolygonList: [],

          // Actions
          // createGmapData: this.createGmapData,
          addNewPolygon: this.addNewPolygon,
          selectPolygon: this.selectPolygon,
          unSelectPolygon: this.unSelectPolygon,
          // generatePolygonList: this.generatePolygonList,

          // Editor actions
          addAction: null,
          subtractAction: this.subtractAction,
          deleteAction: this.deleteAction,
          centerAction: null,

          discardChanges: this.discardChanges,
          persistChanges: this.persistChanges,
        };
      }

      // @TODO make sense to use getDerivedStateFromProps
      componentDidUpdate(prevProps) {
        const { initialGeojson } = this.props;

        if (prevProps.initialGeojson !== initialGeojson) {
          const gmapData = initialGeojson ?
            convertGeojsonToGMData(initialGeojson, this.context) :
            null;
          this.setState({
            gmapData,
            newPolygonList: [],
            polygonList: gmapData ? gmDataToPolygonList(gmapData, this.context) : [],
            selectedPolygonList: [],
          });
        }
      }

      // START Action handlers
      // createGmapData_NOT_IN_USE() {

      // }
      // generatePolygonList_NOT_IN_USE() {
      //   this.setState((state) => {
      //     const polygonList = gmDataToPolygonList(state.gmapData, this.context);
      //     return { polygonList };
      //   });
      // }

      addNewPolygon(polygon) {
        this.setState(state => ({
          newPolygonList: [...state.newPolygonList, polygon],
          selectedPolygonList: [...state.selectedPolygonList, polygon],
        }));
      }

      selectPolygon(polygon) {
        this.setState(state => ({
          selectedPolygonList: [...state.selectedPolygonList, polygon],
        }));
      }
      unSelectPolygon(polygon) {
        this.setState(state => ({
          selectedPolygonList: state.selectedPolygonList.filter(p => p !== polygon),
        }));
      }

      subtractAction(/* ev */) {
        // @TODO convertPolygonToGeoJson
        // @TODO convertPolygonToGeoJsonMulti
        // @TODO convertGeojsonToPolygon
        // @TODO convertGeojsonToPolygonMulti

        convertPolygonToGeoJsonMulti(this.state.selectedPolygonList, this.context)
          .then(geojsonList => subtractPolygons(geojsonList))
          .then(subtractedGeojson => convertGeojsonToPolygon(subtractedGeojson, this.context))
          .then((subtractedPolygon) => {
            this.setState(state => ({
              newPolygonList: [
                ...state.newPolygonList.filter(p => !state.selectedPolygonList.includes(p)),
                subtractedPolygon,
              ],
              polygonList: state.polygonList.filter(p => !state.selectedPolygonList.includes(p)),
              selectedPolygonList: [],
            }));
          });
      }

      deleteAction(/* ev */) {
        this.setState(state => ({
          newPolygonList: state.newPolygonList.filter(p =>
            !state.selectedPolygonList.includes(p)),
          polygonList: state.polygonList.filter(p =>
            !state.selectedPolygonList.includes(p)),
          selectedPolygonList: [],
        }));
      }

      discardChanges(_, cb) {
        this.setState(state => ({
          newPolygonList: [],
          polygonList: state.gmapData ? gmDataToPolygonList(state.gmapData, this.context) : [],
          selectedPolygonList: [],
        }), cb);
      }

      persistChanges(_, cb) {
        // 1. create a new GeoJson<MultiPolygon>
        // 2. send onSave action to the parent, telling it to
        //    update backend with newly created geojson
        createGeoJson([...this.state.newPolygonList, ...this.state.polygonList], this.context)
          .then(({ geojson, totalArea }) => {
            // console.log('[createGeoJson]', geojson, totalArea);

            cb(); // @TODO How do i save new GeoJson and turn off edit mode?
            // @TODO How do i save new GeoJson and turn off edit mode?
            this.props.onSave(geojson, totalArea);
          });
      }
      // END Action handlers

      render() {
        return (
          <PolygonDataContext.Provider value={this.state}>
            {this.props.children}
          </PolygonDataContext.Provider>
        );
      }
}


export default PolygonDataProvider;

PolygonDataProvider.propTypes = {
  children: PropTypes.node.isRequired,
  initialGeojson: GeoJsonType,
  onSave: PropTypes.func,
};

PolygonDataProvider.defaultProps = {
  initialGeojson: null,
  onSave: () => {},
};
