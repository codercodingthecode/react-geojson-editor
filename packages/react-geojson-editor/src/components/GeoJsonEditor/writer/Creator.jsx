import React, { Component } from 'react';

import { GoogleMapContext } from '../google-map';

// @TODO where should I put the code for selecting/unSelecting new polygons ?
// Maybe inside Updator component
export class Creator extends Component {
    static contextType = GoogleMapContext;

    drawingManager = null;

    componentDidMount() {
      const { map, google } = this.context;
      const { onCreate } = this.props;

      this.drawingManager = new google.maps.drawing.DrawingManager({
        drawingMode: google.maps.drawing.OverlayType.POLYGON,
        drawingControl: true,

        drawingControlOptions: {
          position: google.maps.ControlPosition.TOP_CENTER,
          drawingModes: [
            google.maps.drawing.OverlayType.POLYGON,
          ],
        },

        polygonOptions: {
          fillColor: '#455A64',
          fillOpacity: 0.3,
          strokeColor: '#455A64',
          strokeWeight: 3,
          clickable: true,
          editable: true, // @TODO
          zIndex: 1,
        },
      });
      this.drawingManager.addListener('polygoncomplete', (polygon) => {
        polygon.setMap(null);
        onCreate(polygon);
      });

      this.drawingManager.setMap(map);
    }

    componentDidUpdatesaved(/* prevProps */) {
      const { map, google } = this.context;
      const { onCreate } = this.props;

      this.drawingManager = new google.maps.drawing.DrawingManager({
        drawingMode: google.maps.drawing.OverlayType.POLYGON,
        drawingControl: true,

        drawingControlOptions: {
          position: google.maps.ControlPosition.TOP_CENTER,
          drawingModes: [
            google.maps.drawing.OverlayType.POLYGON,
          ],
        },

        polygonOptions: {
          fillColor: '#455A64',
          fillOpacity: 0.3,
          strokeColor: '#455A64',
          strokeWeight: 3,
          clickable: true,
          editable: true, // @TODO
          zIndex: 1,
        },
      });
      this.drawingManager.addListener('polygoncomplete', (polygon) => {
        polygon.setMap(null);
        onCreate(polygon);
      });

      this.drawingManager.setMap(map);
    }


    componentWillUnmount() {
      this.drawingManager.setMap(null);
    }

    render() {
      return null;
    }
}

export default Creator;
