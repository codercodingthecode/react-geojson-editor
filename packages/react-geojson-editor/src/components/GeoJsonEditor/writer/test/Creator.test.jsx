import React from 'react';

import { shallow, mount } from 'enzyme';
import { setMapMock, addListenerMock } from '../../../../../test/helpers/google';

import { GoogleMapContext } from '../../google-map';
import { Creator } from '../Creator';

describe('Button', () => {
  beforeEach(() => {
    setMapMock.mockClear();
    addListenerMock.mockClear();
  });

  it('has not changed', () => {
    const wrapper = shallow(<Creator />);
    expect(wrapper).toMatchSnapshot();
  });
  it('renders by creating drawingManager and attaching it to the map instance', () => {
    // Arrange
    const map = new google.maps.Map();

    // Act
    const wrapper = mount(
      <GoogleMapContext.Provider value={{ map, google }}>
        <Creator />
      </GoogleMapContext.Provider>
    ); 

    // Assert: that gmapData has been attached to the map instance
    expect(wrapper.instance().drawingManager.setMap).toHaveBeenCalledWith(map);
  });
});
