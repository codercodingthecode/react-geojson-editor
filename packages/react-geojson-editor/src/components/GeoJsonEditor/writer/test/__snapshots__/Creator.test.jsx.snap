// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button has not changed 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Creator />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): null,
  Symbol(enzyme.__nodes__): Array [
    null,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
